### PR TITLE
fix: [CDS-83756]: Instrumentation helper code for tracking deployments w.r.t each artifact

### PR DESCRIPTION
**Pull request by vivekkarnatiharness**
## Describe your changes
Have added telemetry code to track artifact deployments for artifacts present in v1 and v2 type of services.
image

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55610)

### CODE REVIEW

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactDeploymentInstrumentationHelper.java
Comment (sarthakkasat) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactDeploymentInstrumentationHelper.java:1]: Can we use the ArtifactSourceInstrumentationHelper class to publish the event?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactDeploymentInstrumentationHelper.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    return sendEvent(eventName, accountId, eventPropertiesMap);
+  }
+
+  public CompletableFuture<Void> sendArtifactDeploymentEvent(ArtifactConfig artifactConfig, String accountId,
+      String orgId, String projectId, String deploymentType, Boolean isServiceV2) {
+    return publishArtifactInfo(
+        artifactConfig, accountId, orgId, projectId, "artifact_deployment", deploymentType, isServiceV2);
+  }
+
+  private CompletableFuture<Void> sendEvent(
```
Comment (sarthakkasat) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactDeploymentInstrumentationHelper.java:62]: Can we not duplicate this method?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/steps/ArtifactsStepV2.java [hunk 2]
```diff
@@ -233,8 +234,14 @@ public AsyncExecutableResponse executeAsyncAfterRbac(
                                   .build());
         taskIds.add(primaryArtifactTaskId);
         artifactConfigMap.put(primaryArtifactTaskId, artifacts.getPrimary().getSpec());
+        artifactDeploymentInstrumentationHelper.sendArtifactDeploymentEvent(artifacts.getPrimary().getSpec(),
+            AmbianceUtils.getAccountId(ambiance), AmbianceUtils.getOrgIdentifier(ambiance),
+            AmbianceUtils.getProjectIdentifier(ambiance), service.getServiceDefinition().getType().name(), true);
       } else if (ACTION.RUN_SYNC.equals(actionForPrimaryArtifact)) {
         artifactConfigMapForNonDelegateTaskTypes.add(artifacts.getPrimary().getSpec());
+        artifactDeploymentInstrumentationHelper.sendArtifactDeploymentEvent(artifacts.getPrimary().getSpec(),
```
Comment (sarthakkasat) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/steps/ArtifactsStepV2.java:242]: Can we call the sendArtifactDeploymentEvent method after the else if block?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/steps/ArtifactsStepV2.java [hunk 3]
```diff
@@ -258,10 +265,16 @@ public AsyncExecutableResponse executeAsyncAfterRbac(
...
+          artifactDeploymentInstrumentationHelper.sendArtifactDeploymentEvent(sidecar.getSidecar().getSpec(),
+              AmbianceUtils.getAccountId(ambiance), AmbianceUtils.getOrgIdentifier(ambiance),
+              AmbianceUtils.getProjectIdentifier(ambiance), service.getServiceDefinition().getType().name(), true);
         } else if (ACTION.RUN_SYNC.equals(actionForSidecar)) {
           checkAndWarnIfDoesNotFollowIdentifierRegex(
               sidecar.getSidecar().getSpec().getIdentifier(), "Sidecar", logCallback);
           artifactConfigMapForNonDelegateTaskTypes.add(sidecar.getSidecar().getSpec());
+          artifactDeploymentInstrumentationHelper.sendArtifactDeploymentEvent(sidecar.getSidecar().getSpec(),
+              AmbianceUtils.getAccountId(ambiance), AmbianceUtils.getOrgIdentifier(ambiance),
+              AmbianceUtils.getProjectIdentifier(ambiance), service.getServiceDefinition().getType().name(), true);
```
Comment (sarthakkasat) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/steps/ArtifactsStepV2.java:277]: Same as above

**Review by sarthakkasat — CHANGES_REQUESTED** [review]

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactDeploymentInstrumentationHelper.java
Comment (vivekkarnatiharness) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactDeploymentInstrumentationHelper.java:1]: Ok, have squashed my changes to this file

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactDeploymentInstrumentationHelper.java (hunk 1 above)
Comment (vivekkarnatiharness) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactDeploymentInstrumentationHelper.java:62]: Done, removed

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/steps/ArtifactsStepV2.java (hunk 2 above)
Comment (vivekkarnatiharness) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/steps/ArtifactsStepV2.java]: Ok, moved it!

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Review by sarthakkasat — DISMISSED** [review]

**Review by abhinavmittal-harness — CHANGES_REQUESTED** [review]
> Reviewed 4 of 6 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55610)* status: 10 file(s) to review. 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactSourceInstrumentationHelper.java` line 55 at r3](https://reviewable.io/reviews/harness/harness-core/55610#-NjqpFTf51_mBaKiXyW6:-NjqpFTf51_mBaKiXyW7:b-pc0tde) ([raw file](https://github.com/harness/harness-core/blob/a661251f4591abd2b84d4e4ae9f099da59073816/125-cd-nextgen/modules/common/core/src/main/java/io/harness/telemetry/helpers/ArtifactSourceInstrumentationHelper.java#L55)):*
> ```Java
>     eventPropertiesMap.put(ARTIFACT_ORG, orgId);
>     eventPropertiesMap.put(ARTIFACT_TYPE, artifactConfig.getSourceType());
>     eventPropertiesMap.put(ARTIFACT_IDENTIFIER, artifactConfig.getIdentifier());
> ```

can we add a ut for checking if all the properties are being added

**Comment by vivekkarnatiharness** [comment]
trigger uts

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55610)* status: 10 file(s) to review. 4. unresolved discussions. 1 of 2 approvals obtained.

**Review by sarthakkasat — APPROVED** [review]

**Comment by vivekkarnatiharness** [comment]
trigger review

**Comment by vivekkarnatiharness** [comment]
trigger messagecheck

**Pull request merged**